### PR TITLE
Introduce a new rule to flag Link in text blocks without `inline`

### PR DESCRIPTION
**Pull request by khiga8**
Relates to: 
* https://github.com/primer/eslint-plugin-primer-react/issues/182
*  https://github.com/github/accessibility/issues/6434

## What

This PR introduces a new accessibility lint rule to ensure that links inside of a text block have the `inline` prop.

## Why

The current recommendation in dotcom is to add `inline` prop on all `Link`s that appear within a text block to ensure that a user’s accessibility link underline setting is respected. 

However, there are currently no automated tools to enforce that developers actually set this prop on their links. As a result, developers continue to add Links to text blocks contributing to [1.4.1: Use of Color accessibility issues](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html).

By having this lint rule in place, we can proactively surface to developers when they should set the `inline` prop.

## Implementation details

The logic essentially treats any `Link` that appears with a string literal on either side of it as a link within a text block. It checks for the presence of `inline` or `inline={true}`. 

For example, the following will be flagged:

```.tsx
<p>
  Learn more about <Link href={...}>GitHub</Link>.
</p>
```

```.tsx
<SomeComponent>
  <Link href={...}>GitHub</Link> is where you share code.
</SomeComponent>
```

In addition, this lint rule aims to minimize false positives and will skip Links even if adjacent text is detected for the following scenarios:

* `<Link sx={{fontWeight:...}}>` or `<Link sx={{fontFamily:...}}>` - these technically may provide sufficient distinguishing styling.
* `<Link>` where the only adjacent text is a period -  these can't really be considered a text block.
* `<Link>` where the children is a JSX component, rather than a string literal -  this is because the JSX component may be an icon link, in which case it may be sufficient to distinguish.
* `<Link>` that are nested inside of headings -  these feel safer to skip...

## Testing / Validation

I've tested this branch against dotcom to make sure there's nothing unexpected! This flagged around 140 `Link` missing inline in dotcom. I've gone ahead and opened fixes for whatever was flagged by this rule [across 8 PRs](https://github.com/github/github/pulls?q=is%3Apr+is%3Aopen+%22%5Breact-link-inline+%22+author%3Akhiga8).

I broke this up into reviewable chunks so that codeowners can make sure to minimize false positives raised by this rule. (Related: https://github.com/github/accessibility/issues/6434)

## Reviewers 📣

1. I am open to any feedback on this logic! Is there anything I'm missing?
2. Is it okay to enable this rule in `recommended.js` or is this only relevant to dotcom?
3. Right now this rule will flag `<Link underline>` even though technically `underline` does provide sufficient styling. I decided to flag `underline` because I noticed it is deprecated so maybe this is an opportunity for consumers to migrate to `inline`. Let me know if you think we shouldn't flag `underline`!

### CODE REVIEW

File: src/rules/__tests__/a11y-link-in-text-block.test.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    <Box>
+
+    <Link href="something">
+      Blah blah
+    </Link>{' '}
+    .
+    </Box>
+`,
+  ],
+  invalid: [
```
Comment (khiga8) [inline src/rules/__tests__/a11y-link-in-text-block.test.js:102]: TODO: Add test case for how it will flag `<Link underline>` within a text block.

**Comment by khiga8** [comment]
I'm still in the process of getting feedback to make sure there's nothing unexpected with the instances flagged by this rule via [PRs in dotcom](https://github.com/github/github/pulls?q=%22%5Breact-link-inline%22+author%3Akhiga8) but want to open this up for early review!

File: .changeset/happy-parents-invite.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"eslint-plugin-primer-react": patch
+---
+
+New rule to flag links in text block missing the `inline` prop
```
Comment (siddharthkp) [inline .changeset/happy-parents-invite.md:5]:
```suggestion
New rule to flag `Link` in text block missing the `inline` prop
```

**Comment by siddharthkp** [comment]
Makes sense to me on a theoretical level, hard to review without testing it. Can we run this once on gh/gh to see if it catches any false positives?

**Comment by khiga8** [comment]
All [PRs in dotcom](https://github.com/github/github/pulls?q=%22%5Breact-link-inline%22+author%3Akhiga8) have been reviewed and merged without any issues! 
This is ready for a formal review! :D

File: src/configs/recommended.js [hunk 3]
```diff
@@ -16,6 +16,7 @@ module.exports = {
     'primer-react/new-color-css-vars': 'error',
     'primer-react/a11y-explicit-heading': 'error',
     'primer-react/no-deprecated-props': 'warn',
+    'primer-react/a11y-link-in-text-block': 'error',
```
Comment (khiga8) [inline src/configs/recommended.js:19]: Given that `inline` is specific to dotcom's setup, maybe this should not be turned on by default. would love to hear what others think!

Comment (siddharthkp) [reply src/configs/recommended.js:19]: True, we shouldn't turn it on by default

**Review by siddharthkp — APPROVED** [review]
LGTM!

File: src/configs/recommended.js (hunk 3 above)
Comment (khiga8) [inline src/configs/recommended.js:19]:
```suggestion
```

**Pull request merged**